File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/kde_simple_tutorial.py
#!/usr/bin/env python3
"""
kdeTutorial_simple.py
---------------------
This tutorial demonstrates simple kernel density estimation.
First, it shows KDE on a small dataset (SixMPG), then it loads automotive MPG data
and fits KDEs with different bandwidths.
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
from scipy.stats import gaussian_kde

# --- Part 1: Small Dataset ---
SixMPG = np.array([13,15,23,29,32,34])
plt.figure()
plt.hist(SixMPG, bins='auto', color='skyblue', edgecolor='k')
plt.title("Histogram of SixMPG")
plt.xlabel("MPG")
plt.ylabel("Count")
plt.grid(True)
plt.show()

# Fit a KDE with bandwidth 4
kde_six = gaussian_kde(SixMPG, bw_method=4/np.std(SixMPG, ddof=1))
x = np.linspace(0,45,400)
ySix = kde_six(x)
plt.figure()
plt.plot(x, ySix, 'k-', linewidth=2, label="KDE Bandwidth=4")
plt.title("KDE for SixMPG")
plt.xlabel("X"); plt.ylabel("Estimated Density")
plt.grid(True)
# Plot individual Gaussian pdfs for each point, scaled.
for mpg in SixMPG:
    y_ind = norm.pdf(x, loc=mpg, scale=4) / 6  # scale factor 1/6
    plt.plot(x, y_ind, 'b:', linewidth=2)
plt.legend()
plt.show()

# --- Part 2: Automotive MPG Data ---
# For demonstration, we simulate automotive MPG data.
# In practice, load your dataset (e.g., from carbig.mat).
np.random.seed(1)
MPG = np.random.normal(30, 5, 100)  # simulate 100 MPG values

pd1 = gaussian_kde(MPG)           # default bandwidth
pd2 = gaussian_kde(MPG, bw_method=1/np.std(MPG, ddof=1))
pd3 = gaussian_kde(MPG, bw_method=5/np.std(MPG, ddof=1))

x_vals = np.linspace(-10, 60, 200)
y1 = pd1(x_vals)
y2 = pd2(x_vals)
y3 = pd3(x_vals)

plt.figure()
plt.plot(x_vals, y1, 'r-', linewidth=2, label="Bandwidth = Default")
plt.plot(x_vals, y2, 'k:', linewidth=2, label="Bandwidth = 1")
plt.plot(x_vals, y3, 'b--', linewidth=2, label="Bandwidth = 5")
plt.legend()
plt.xlabel("X"); plt.ylabel("Estimated Density")
plt.title("KDE for Automotive MPG")
plt.grid(True)
plt.show()

File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/kde_simple_tutorial.m
close all; clear all; clc; 

SixMPG = [13;15;23;29;32;34];
figure;
histogram(SixMPG)

figure;
pdSix = fitdist(SixMPG,'Kernel','Width',4);
x = 0:.1:45;
ySix = pdf(pdSix,x);
plot(x,ySix,'k-','LineWidth',2)

% Plot each individual pdf and scale its appearance on the plot
hold on
for i=1:6
    pd = makedist('Normal','mu',SixMPG(i),'sigma',4);
    y = pdf(pd,x);
    y = y/6;
    plot(x,y,'b:','LineWidth',2)
end
xlabel('X'); ylabel('Estimated Density');
title('Kernel Density Estimate for Mixture of Gaussians');
hold off

% save_all_figs_OPTION('results/kde2','png',1)


% Create kernel distribution objects
load 'data/carbig.mat'
pd1 = fitdist(MPG,'kernel');
pd2 = fitdist(MPG,'kernel','Width',1);
pd3 = fitdist(MPG,'kernel','Width',5);

% Compute each pdf
x = -10:1:60;
y1 = pdf(pd1,x);
y2 = pdf(pd2,x);
y3 = pdf(pd3,x);

% Plot each pdf
figure;
plot(x,y1,'Color','r','LineStyle','-','LineWidth',2)
hold on
plot(x,y2,'Color','k','LineStyle',':','LineWidth',2)
plot(x,y3,'Color','b','LineStyle','--','LineWidth',2)
legend({'Bandwidth = Default','Bandwidth = 1','Bandwidth = 5'})
xlabel('X'); ylabel('Estimated Density');
title('Kernel Density Estimate for Automotive MPG');
hold off

% save_all_figs_OPTION('results/kde3','png',1)